spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/AbstractBinder.java

```diff
@@ -538,31 +538,31 @@ protected RetryTemplate buildRetryTemplateIfRetryEnabled(DefaultBindingPropertie
 
 	/**
 	 * Handles representing any java class as a {@link MimeType}.
+	 *
 	 * @author David Turanski
-	 * @see <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Class.html#getName"/>
+	 * @author Ilayaperumal Gopinathan
 	 */
 	abstract static class JavaClassMimeTypeConversion {
 
-		public static final MimeType APPLICATION_OCTET_STREAM_MIME_TYPE = MimeType.valueOf
-				(APPLICATION_OCTET_STREAM_VALUE);
-
-		public static final MimeType TEXT_PLAIN_MIME_TYPE = MimeType.valueOf(TEXT_PLAIN_VALUE);
-
 		private static ConcurrentMap<String, MimeType> mimeTypesCache = new ConcurrentHashMap<>();
 
-		static MimeType mimeTypeFromObject(Object obj) {
-			Assert.notNull(obj, "object cannot be null.");
-			if (obj instanceof byte[]) {
-				return APPLICATION_OCTET_STREAM_MIME_TYPE;
+		static MimeType mimeTypeFromObject(Object payload, Object originalContentType) {
+			Assert.notNull(payload, "payload object cannot be null.");
+			if (payload instanceof byte[]) {
+				return MimeTypeUtils.APPLICATION_OCTET_STREAM;
 			}
-			if (obj instanceof String) {
-				return TEXT_PLAIN_MIME_TYPE;
+			if (payload instanceof String) {
+				if (originalContentType != null &&
+						originalContentType.toString().equals(MimeTypeUtils.APPLICATION_JSON_VALUE)) {
```

Comment:
Trying to understand why this is necessary. Since the original content type will replace the CONTENT_TYPE header in the deserialized message, the only function of the `text/plain` content type header is to signal that the payload must be deserialized as a String, before the header is replaced with the original `application/json` value.